app/core/model/src/main/java/io/syndesis/model/metrics/IntegrationDeploymentMetrics.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (C) 2016 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.syndesis.model.metrics;
+
+import java.io.Serializable;
+import java.util.Date;
+import java.util.Optional;
+
+import org.immutables.value.Value;
+
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
+@Value.Immutable
+@JsonDeserialize(builder = IntegrationDeploymentMetrics.Builder.class)
+@SuppressWarnings("immutables")
+public interface IntegrationDeploymentMetrics extends Serializable{
```

Comment:
Why do you need an new class which looks exactly the same as IntegrationMetrics propertywise ? Question is, would this ever differ ?

Actually for the response of the RestEndpoint I think a metrics structure like

```json
{ 
    "total": {
        "nrMessage": 20,
        ....
    },

    "deployments": [{
         "nrMessages": 2,
         ...
     }, {
         "nrMessages": 20,
         ....
     },....
     ]
}
```

would allow for having only a single 'metrics' object which only holds the values. So, not a nested, duplicated structure, but a flat one.

Regardless of this, I wonder how the correlation to the concrete Deployment is ? Is it by order in the array ? I would add somewhere the deployment id, to make it easier for the UI to know which deployment metrics belongs to which deployment. Also adding the integration id in the response helps the UI for correlation so that it does not require to keep the request along with the response, but the response stands for it alone.

About naming:(and i'm nearly sure you will ignore this, but nevertheless): I would call it `getNrMessages()` and not `getMessages()` as this method does not return messages. Same for `getLastProcessed()` which does not return the last processed message.

Reply:
the objects are 'withId' so it holds either the integrationId or the deploymentId. I prefer a hierarchical structure as this how consumers typically use it.

Reply:
@rhuss 

1. I removed the second commit, and went back to the original. What I have there is very much the same as what you are proposing. So let's talk tomorrow.

2. You requested it to be named 'messages' before. You are a hard one to please :)

Reply:
Right, I suggested to change `numberOfProcessedMessages` to be shorter, e.g. `messages` in #1156. And right, from time to time I evolve my opinion (and even sometimes revert it, but not here), too, as the new suggestion was indeed shorter but suffers the same semantic issue which I haven't seen at that time. An no, I'm not so hard to please, just picky about naming ;-) Happy to discuss about API naming, too, as I think its important, though often overlooked.

Again, why is having the following structure so bad, to avoid duplication and recursive definitions:

```java
class Metrics {
    int nrMessages;
    int nrErrors;
    Date start;
    Date lastProcessed;
}

class IntegrationMetrics {
      
    Metrics summary;  // or just: "metrics"
    
     List<Metrics> deploymentMetrics; // or 'integrationDeploymentMetrics'
}
```

This structure is also easy to evolve, so when you have to add a new datum you just add it to `Metrics`.  If both kind of metrics are likely to evolve differently, then indeed two `Metric` might make sense, but we can still add this when this is the case.

@syndesisio/backend @syndesisio/ui-api Any opinions on this on ? E.g. what is the easiest way to consume metrics for an integration overall (across all deployments) and for each individual deployments. Its about the structure coming back for a request for the metrics for a certain integration.

Reply:
@rhuss I was trying to not break backwards compatibility until things are working between the teams that's all.